dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/DynamicPureLoopCutting.java

```diff
@@ -0,0 +1,381 @@
+    private List<Loop> findLoops(Thread thread) {
+        //NOTE: This needs to get updated if the unrolling code renames the loop labels differently
+        final String iterNumberSeparator = "itr_";
+
+        int loopCounter = 0;
+        final Map<String, Loop> labelName2LoopMap = new HashMap<>();
+        final List<Loop> loops = new ArrayList<>();
+        for (Event e : thread.getEvents()) {
+            if (!(e instanceof Label && ((Label)e).getName().lastIndexOf(iterNumberSeparator) >= 0)) {
+                // No loop label
+                continue;
+            }
+            final Label loopLabel = (Label) e;
+            final int iterNumberSeparatorIndex = loopLabel.getName().lastIndexOf(iterNumberSeparator);
+            final int iterNumber = Integer.parseInt(loopLabel.getName()
+                    .substring(iterNumberSeparatorIndex + iterNumberSeparator.length()));
+            final String loopName = loopLabel.getName().substring(0, iterNumberSeparatorIndex);
+
+            if (iterNumber == 1) {
+                final Loop loop = new Loop();
+                loop.thread = thread;
+                loop.loopNumber = loopCounter++;
+                labelName2LoopMap.put(loopName, loop);
+                loops.add(loop);
+            }
+            final Loop loop = labelName2LoopMap.get(loopName);
+            Verify.verifyNotNull(loop,
+                    "Found intermediate loop iteration {} before encountering loop beginning.", loopLabel.getName());
+            final LoopIteration iter = new LoopIteration();
+            iter.loop = loop;
+            iter.iterationNumber = iterNumber;
+            iter.startLabel = loopLabel;
+            if (!loop.iterations.isEmpty()) {
+                loop.iterations.get(loop.iterations.size() - 1).endLabel = iter.startLabel;
+            }
+            loop.iterations.add(iter);
+        }
+
+        loops.forEach(this::populateLoopSideEffectsAndExitPoints);
+        return loops;
+    }
+
+    private void populateLoopSideEffectsAndExitPoints(Loop loop) {
+        loop.iterations.subList(0, loop.iterations.size() - 1).forEach(this::populateLoopIteration);
+    }
+    private void populateLoopIteration(LoopIteration iteration) {
+        Preconditions.checkNotNull(iteration.endLabel);
+        final Label start = iteration.startLabel;
+        final Label end = iteration.endLabel;
+
+        iteration.sideEffects.addAll(collectSideEffects(start, end));
+
+        Event cur = start;
+        while ((cur = cur.getSuccessor()) != end) {
+            if (cur instanceof CondJump) {
+                final CondJump jump = (CondJump) cur;
+                final Label jumpTarget = jump.getLabel();
+                if (jump.isGoto() && jumpTarget.getCId() > end.getCId()) {
+                    iteration.trueExitPoints.add(jump);
+                }
+            }
+        }
+    }
+
+    private List<Event> collectSideEffects(Label iterStart, Label iterEnd) {
```

Comment:
Seems repeated code from another pass. Either refactor or add a TODO to get track of the potential refactoring

Reply:
It is not identical code. The other one terminates on finding the first side effect, this one does not.
The intention is slightly different.